src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -845,18 +849,317 @@ public Document GetRevisionBefore(DocumentsOperationContext context, string id,
+        }
+
+        public async Task<IOperationResult> RevertRevisions(DateTime before, TimeSpan window, Action<IOperationProgress> onProgressAction = null)
+        {
+            var list = new List<Document>();
+            var result = new RevertResult();
+
+            var parameters = new RevertParameters
+            {
+                Before = before,
+                MinimalDate = before.Add(-window), // since the documents/revisions are not sorted by date, stop searching if we reached this date.
+                EtagBarrier = _documentsStorage.GenerateNextEtag() // every change after this etag, will _not_ be reverted.
+            };
+            parameters.LastScannedEtag = parameters.EtagBarrier;
+
+            var hasMore = true;
+            while (hasMore)
+            {
+                using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext writeCtx))
+                {
+                    hasMore = PrepareRevertedRevisions(writeCtx, parameters, result, list);
+                    await WriteRevertedRevisions(list);
+                    onProgressAction?.Invoke(result.Progress);
+                    result.Progress.Warnings.Clear();
+                }
+            }
+
+            return result;
+        }
+
+        private async Task WriteRevertedRevisions(List<Document> list)
+        {
+            if (list.Count == 0)
+                return;
+
+            await _database.TxMerger.Enqueue(new RevertDocumentsCommand(list));
+
+            list.Clear();
+        }
+
+        private bool PrepareRevertedRevisions(DocumentsOperationContext writeCtx, RevertParameters parameters, RevertResult result, List<Document> list)
+        {
+            using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext readCtx))
+            using (readCtx.OpenReadTransaction())
+            {
+                var revisions = new Table(RevisionsSchema, readCtx.Transaction.InnerTransaction);
+                foreach (var tvr in revisions.SeekBackwardFrom(RevisionsSchema.FixedSizeIndexes[AllRevisionsEtagsSlice],
+                    parameters.LastScannedEtag))
+                {
+                    result.Progress.ScannedRevisions++;
+
+                    var id = TableValueToString(readCtx, (int)RevisionsTable.LowerId, ref tvr.Reader);
+                    var etag = TableValueToEtag((int)RevisionsTable.Etag, ref tvr.Reader);
+                    parameters.LastScannedEtag = etag;
+
+                    if (parameters.ScannedIds.Add(id) == false)
+                        continue;
+
+                    result.Progress.ScannedDocuments++;
+
+                    if (etag > parameters.EtagBarrier)
+                    {
+                        result.Progress.Warn(id, "This document wouldn't be reverted, because it changed after the revert progress started.");
+                        continue;
```

Comment:
Shouldn't this be a `break` ?

Reply:
This of the following timeline:
1. We request the revert at local etag of 100
2. Some inserted several documents and increase the local etag to 110. 
3. We start iteration over the revisions from the top. (from etag 110)

So we want to skip those document that was changed after etag 100, but we want to revert the rest.